main.go

```diff
@@ -53,36 +55,68 @@ func main() {
 		return
 	}
 
-	err = validation.Validate(opts)
-	if err != nil {
-		logger.Printf("%s", err)
-		os.Exit(1)
+	if err = validation.Validate(opts); err != nil {
+		logger.Fatalf("%s", err)
 	}
 
 	validator := NewValidator(opts.EmailDomains, opts.AuthenticatedEmailsFile)
 	oauthproxy, err := NewOAuthProxy(opts, validator)
 	if err != nil {
-		logger.Errorf("ERROR: Failed to initialise OAuth2 Proxy: %v", err)
-		os.Exit(1)
+		logger.Fatalf("ERROR: Failed to initialise OAuth2 Proxy: %v", err)
 	}
 
 	rand.Seed(time.Now().UnixNano())
 
+	oauthProxyStop := make(chan struct{}, 1)
+	metricsStop := startMetricsServer(opts.MetricsAddress, oauthProxyStop)
+
 	s := &Server{
 		Handler: oauthproxy,
 		Opts:    opts,
-		stop:    make(chan struct{}, 1),
+		stop:    oauthProxyStop,
 	}
 	// Observe signals in background goroutine.
 	go func() {
 		sigint := make(chan os.Signal, 1)
 		signal.Notify(sigint, os.Interrupt, syscall.SIGTERM)
 		<-sigint
-		s.stop <- struct{}{} // notify having caught signal
+		s.stop <- struct{}{} // notify having caught signal stop oauthproxy
+		close(metricsStop)   // and the metrics endpoint
 	}()
 	s.ListenAndServe()
 }
 
+// startMetricsServer will start the metrics server on the specified address.
+// It always return a channel to signal stop even when it does not run.
+func startMetricsServer(address string, oauthProxyStop chan struct{}) chan struct{} {
+	stop := make(chan struct{}, 1)
+
+	// Attempt to setup the metrics endpoint if we have an address
+	if address != "" {
+		s := &http.Server{Addr: address, Handler: middleware.DefaultMetricsHandler}
```

Comment:
If a user configures TLS for the main server -- I would think we would want the metrics endpoint to listen with the same `tls.Config` right?

Reply:
I think you might actually want to have a second TLS config for the metrics, for example, metrics are likely to be internal so you could just have a self signed cert for metrics for example

For now I might suggest we keep metrics as HTTP only and investigate adding HTTPs support later